myhpom/tests/test_cloudfactory_response_view.py

```diff
@@ -0,0 +1,72 @@
+import json
+import os
+from django.core.urlresolvers import reverse
+from django.test import TestCase
+
+from myhpom.models import CloudFactoryDocumentRun
+
+CF_PATH = os.path.join(os.path.dirname(__file__), 'fixtures', 'cloudfactory')
+SUCCESS_DATA = open(os.path.join(CF_PATH, 'callback_success.json')).read()
+
+
+class CloudfactoryResponseTest(TestCase):
+
+    def test_no_data(self):
+        # When no data is posted to the viewpoint, a 4xx code is returned,
+        # alerting the requestor that the problem is theirs
+        response = self.client.post(reverse('myhpom:cloudfactory_response'))
+        self.assertEqual(400, response.status_code)
+
+    def test_badly_formed_data(self):
+        # When non-json data is sent, we return a 400 to alert the user to the
+        # bad formatting.
+        response = self.client.post(
+            reverse('myhpom:cloudfactory_response'), data='blah', content_type='text/example')
+        self.assertEqual(400, response.status_code)
+
+    def test_no_id(self):
+        # Record doesn't exist on our end - we want a log of this.
+        response = self.client.post(reverse('myhpom:cloudfactory_response'), data=SUCCESS_DATA, content_type='application/json')
+        self.assertEqual(404, response.status_code)
+
+    def test_already_finished(self):
+        # When CloudFactoryDocumentRun exists, but it is already 'finished' - we
+        # don't expect to receive any updates
+        run = CloudFactoryDocumentRun.objects.create(run_id='E7gVrtVQJx')
+        for status in CloudFactoryDocumentRun.STATUS_FINAL_STATES:
+            run.status = status
+            run.save()
+            response = self.client.post(reverse('myhpom:cloudfactory_response'), data=SUCCESS_DATA, content_type='application/json')
+            self.assertEqual(400, response.status_code, '%s should fail' % status)
+
+    def test_failed_review(self):
+        # We get back a response that says any of the output criteria failed
+        run = CloudFactoryDocumentRun.objects.create(run_id='E7gVrtVQJx')
+
+        # Try all the possible keys that could fail - they should all cause the
+        # run to transition to finished.
+        failed_data = json.loads(SUCCESS_DATA)
+        for output in failed_data['units'][0]['output'].keys():
+            failed_data['units'][0]['output'][output] = False
```

Comment:
I'm not confident that this test isolates _each_ key - the previous ones will still be false (python block scope). For example:
```
>>> d = {'a': True, 'b': True, 'c': True}
>>> for k in d:
...     d[k] = False; print d
... 
{'a': False, 'c': True, 'b': True}
{'a': False, 'c': False, 'b': True}
{'a': False, 'c': False, 'b': False}
>>> 
```

Reply:
Woops, I'll move the failed_data inside the loop.